website/addons/box/model.py

```diff
@@ -0,0 +1,450 @@
+# -*- coding: utf-8 -*-
+import os
+import time
+import logging
+import hashlib
+from datetime import datetime
+
+import furl
+import requests
+from box import CredentialsV2
+from modularodm import fields, Q, StoredObject
+from modularodm.exceptions import ModularOdmException
+
+from framework.auth import Auth
+from website.addons.base import exceptions
+from website.addons.base import AddonUserSettingsBase, AddonNodeSettingsBase, GuidFile
+
+from website.addons.box import settings
+from website.addons.box.utils import BoxNodeLogger
+from website.addons.box.client import get_client_from_user_settings
+
+
+logger = logging.getLogger(__name__)
+
+
+class BoxFile(GuidFile):
+    """A Box file model with a GUID. Created lazily upon viewing a
+    file's detail page.
+    """
+
+    #: Full path to the file, e.g. 'My Pictures/foo.png'
+    path = fields.StringField(required=True, index=True)
+
+    @property
+    def waterbutler_path(self):
+        if not self.path.startswith('/'):
+            return '/{}'.format(self.path)
+        return self.path
+
+    @property
+    def provider(self):
+        return 'box'
+
+    @property
+    def version_identifier(self):
+        return 'revision'
+
+    @property
+    def unique_identifier(self):
+        # TODO
+        return hashlib.md5(self.waterbutler_path).hexdigest()
+
+    @classmethod
+    def get_or_create(cls, node, path):
+        """Get or create a new file record. Return a tuple of the form (obj, created)
+        """
+        try:
+            new = cls.find_one(
+                Q('node', 'eq', node) &
+                Q('path', 'eq', path)
+            )
+            created = False
+        except ModularOdmException:
+            # Create new
+            new = cls(node=node, path=path)
+            new.save()
+            created = True
+        return new, created
+
+
+class BoxOAuthSettings(StoredObject):
+    """
+    this model address the problem if we have two osf user link
+    to the same box user and their access token conflicts issue
+    """
+
+    # Box user id, for example, "4974056"
+    user_id = fields.StringField(primary=True, required=True)
+    # Box user name this is the user's login
+    username = fields.StringField()
+    _access_token = fields.StringField()
+    refresh_token = fields.StringField()
+    expires_at = fields.DateTimeField()
+
+    @property
```

Comment:
Given that this method may make an HTTP request (and therefore raise an HTTP error) and that  properties have strange error handling behavior, we might want to make this a method (maybe `fetch_access_token`.

That way, we could also rename `_access_token` to `access_token` and get rid of the setter.

This does not necessarily need to happen before this is merged.